### ISSUE TITLE
Decide on code publicity/license

### ISSUE THREAD

**Issue by finnnnnnnnnnnnnnnnn** [HellFire-Creators/HellFire-Site#8]
We should probably decide on what the license of the code should be as well as if we should make the repo public.

**Comment by Conorsmine** [comment]
That sounds wonderful.
However, how do we want to provide access to the API then? API keys probably. The issue is, how do we distinguish between malicious and meaningful intent?

**Comment by some16** [comment]
I'm not totally sure what you mean by this, since the worst thing the user could do is to remove all the files from the project, but they can't delete the repository since its owned by the bot, meaning that another user can just undo that change since we are using Github.

It's possible you're referring to maybe how API keys should be given out, in which case probably the best way to handle this is by having a feature of the site that gives the user the ability to generate keys.

**Comment by Conorsmine** [comment]
What I am worrying about is, how we figure out if someone is a part of a project or not.
What I could imagine is, that we generate an access token, or API key, for that specific project. So everyone can use the API, but only those with the access key can actually read and write to it.
Otherwise, people could create a simple script to delete all files in every creator's repository.
But then how do we find out who is part of the project if we allow 3rd party access to the API? Do they still have to register the project on our website to get the key AND then use it with someone else's frontend?

**Comment by some16** [comment]
What if we just generated an access token for the user instead of the project? They would have to login to our website to generate the token once, but they only have to do this once and then they can use it to access any project they have permissions for.

**Comment by Conorsmine** [comment]
I have another suggestion.
We have two systems. One for the creators and another for access to the API:

### Side note
A small side note, since I'm going to bring this up in a second.
Creators have a UUID. Something we can "share" freely without it being any danger to the security of the creator.
This UUID is unique, as the name would suggest, to a creator and will always be theirs. They may change their name or other account details, but this will always stay the same. This will also then be used as the primary key in the database for querying creators.

## Creators
--
Every creator has one single auth key, this key is basically their permission to read & write to their projects.
Creators may share this key with 3rd party apps. So maybe someone created an alternative version of the website, some might like it more. They share their key with that app.
The app sends requests to the API; The UUID will tell us which creator they are talking about and the auth key of the creator tells us if it is actually them or not.

## APIs
-- 
In my opinion, developers should always need an API key.
This is mainly for two reasons:
- We can log requests and check for malicious/irresponsible requests to the API; We can then use the API key to distinguish between different API users
- We can and this might sound harsh, "enforce" rules upon API users. These should always only be something like: "Have your projects open source under a GNU license." So we and others can actually inspect their code and look for any suspicious activity.

Moreover, it would mean that API users would need to come to us and ask for an API key.
So we as devs get into contact with the API users, I find this very important. It would be great to get feedback from them, but also for maybe integrating them into the dev team in the future, if we so choose.

**Comment by some16** [comment]
So this is quite similar to how I was planning it, the main difference being there isn't really a need for a uuid to be ever interacted with by the user or even the 3rd party apps, since the token functions as the identifier, as well as the fact that in the database if a user decides to change there username or something its fine since Django handles users so this isn't really something that has to be accounted for.

I'm not quite sure by what you mean by developers in this case, unless your saying those who develop third party apps and you are saying that they also need a token to be sent along with the users token? This is probably a good idea.

**Comment by Conorsmine** [comment]
Correct, I wanted something we could identify 3rd party apps by. So in an "emergency" we can block their requests.

I wanted the UUID, so we can have something like a "HellFire User" in the DB.
This "HellFire User" entry will be the bridge between Minecraft, discord and creator accounts.
Moreover, I wanted the UUID so we can refer to users and don't have to keep the UUID a secret. If we only use the auth token for creators to identify them, this could lead to possible security risks, as accidentally exposing it, would lead to access to all the creators' projects.
Therefore, I assumed it would be beneficial to have a non-critical way of referring to a creator via a UUID and then only use the auth key to check for permissions.

So we could have API requests, like returning all the projects a specific creator is part of. I would find it overkill to require the auth key for that.

So to sum up:
I'd have a UUID for every creator as a non-crucial way of identifying. This could allow us to have some more API possibilities without having to require the auth key; A very crucial secret for the creator.
And then only use the auth key for actual read and write requests.

**Comment by Conorsmine** [comment]
Also, I wasn't very clear with my use:
With "API users", I was referring to the developers using the API.

**Comment by some16** [comment]
I agree with you I think that makes sense, there is one thing though about this
 > So we could have API requests, like returning all the projects a specific creator is part of. I would find it overkill to require the auth key for that.

I assume this would only be public projects the creator is a part of it would return? Or should the existence of private projects be public?

**Comment by Conorsmine** [comment]
I rather had in mind, that projects have "statuses", like "complete", etc.
And that only "complete"/"published" projects would then be returned. This would also have the benefit that projects could automatically be published in the discord server as well.
But I would never send data that the creators haven't explicitly allowed to be shared.
